### PR TITLE
feat: mongo transactions

### PR DESCRIPTION
**Pull request by niall-shaw**
Solves the same problem as https://github.com/didx-xyz/ssi-trust-registry/pull/197, but by using transactions instead of manual deletions / updates.
This has the added benefit of not exposing any unnecessary `delete` functions, but requires some extra overhead of creating and passing a mongo `session`.

Sessions and transactions require mongodb [replica sets](https://www.mongodb.com/docs/manual/replication/), which is why there were several changes required to be made to the docker-compose.

I believe this is the better/correct approach, as production uses replica sets anyway.

Unfortunately, it isn't possible to start the default github actions' mongodb service with replica sets, hence the use of the third party action.

Working with docker-compose, tilt, local development (i.e. `yarn dev`) and CI.

**Note: For local development, or environments using single node replica sets, you must include `?directConnection=true` as part of your connection string.**

### CODE REVIEW

File: packages/backend/src/submission/mongoRepository.ts [hunk 1]
```diff
@@ -64,20 +65,26 @@ async function addSubmission(collection: Collection, submission: Submission) {
...
-  id: string,
-  state: 'approved' | 'rejected',
+  submission: Submission,
+  config: { session?: ClientSession; newDatestamps?: boolean } = {
+    newDatestamps: true,
+  },
 ) {
-  const submission = await collection.findOneAndUpdate(
+  const { id, ...data } = submission
+  if (config.newDatestamps) {
```
Comment (jakubkoci) [inline packages/backend/src/submission/mongoRepository.ts:76]: What's the purpose of `newDatestamps`? Shouldn't we update the date every time we do the update? 🤔

File: packages/backend/src/entity/mongoRepository.ts [hunk 2]
```diff
@@ -43,11 +43,15 @@ async function findByDid(collection: Collection, did: string) {
   return result && Entity.parse(result)
 }
 
-async function addEntity(collection: Collection, entity: Entity) {
+async function addEntity(
+  collection: Collection,
+  entity: Entity,
+  config?: { session?: ClientSession },
```
Comment (jakubkoci) [inline packages/backend/src/entity/mongoRepository.ts:49]: I'm not sure about using `config` as the name of a function parameter in this case. What's the meaning of that? Is it a config of an entity, app, or something else? What if we use just `session`?

File: packages/backend/src/submission/mongoRepository.ts (hunk 1 above)
Comment (niall-shaw) [reply packages/backend/src/submission/mongoRepository.ts:76]: `newDatestamps` was a property only relevant to the manual revert transaction approach, i.e. to set the `updatedAt` back to what it was before the failed update transaction.

Accidentally copied over to this branch, will remove.

File: packages/backend/src/entity/mongoRepository.ts (hunk 2 above)
Comment (niall-shaw) [reply packages/backend/src/entity/mongoRepository.ts:49]: `session` is fine, now that you pointed out that `newDatestamps` isn't necessary in this PR.
`config` was just an attempt to structure the data if there were multiple params.

File: packages/frontend/src/common/hooks.ts [hunk 3]
```diff
@@ -0,0 +1,12 @@
+'use client'
+
+import { FieldValues, UseFormProps, useForm } from 'react-hook-form'
+
+export function useFormWithServerError<
```
[the remark this answers is not in the document]
Comment (niall-shaw) [reply packages/frontend/src/common/hooks.ts:5]: Allows the usage of `setError('server')` and adds the `server` property to the `errors` object, without the need to repetitively add `server?: never` to the inputs object.

[the remark this answers is not in the document]
Comment (niall-shaw) [reply packages/frontend/src/common/hooks.ts:5]: i.e. reusable generic server errors

File: packages/backend/src/email/client-stub.ts [hunk 4]
```diff
@@ -13,21 +13,30 @@ export interface EmailClientStub {
...
     async sendMailFromTemplate(
       to: string,
       subject: string,
       templatePath: string,
       templateParams: Record<string, unknown>,
     ) {
+      if (shouldFailNextSend) {
+        shouldFailNextSend = false
+        throw new Error('Failed to send email')
+      }
```
[the remark this answers is not in the document]
Comment (niall-shaw) [reply packages/backend/src/email/client-stub.ts:33]: I was thinking the same. Does that mean we should remove the stub and replace with a mock? 
Or just create a mock for these few cases?

[the remark this answers is not in the document]
Comment (niall-shaw) [reply packages/backend/src/email/client-stub.ts]: For now, I created a mock just for use in a subset of tests - `transaction rollbacks`

File: packages/frontend/src/common/hooks.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (jakubkoci) [reply packages/frontend/src/common/hooks.ts:5]: Could you please write down a few examples of how would the repetitive code look like?

Custom hooks bring a lot of complexity, so I would rather avoid them if possible.

File: .docker/mongodb/scripts/rs-init.js
Comment (jakubkoci) [inline .docker/mongodb/scripts/rs-init.js:1]: What does the `rs` in the name mean? 🤔

File: packages/backend/src/email/client-stub.ts (hunk 4 above)
[the remark this answers is not in the document]
Comment (jakubkoci) [reply packages/backend/src/email/client-stub.ts]: Yes, that's the correct approach I think :)

File: .docker/mongodb/initdb.d/usercreate.sh [hunk 5]
```diff
@@ -0,0 +1,9 @@
+set -Eeuo pipefail
+ 
+mongosh -u "$MONGO_INITDB_ROOT_USERNAME" -p "$MONGO_INITDB_ROOT_PASSWORD" --authenticationDatabase admin "$MONGO_INITDB_DATABASE" <<EOF
+    db.createUser({
+        user: '$MONGO_INITDB_ROOT_USERNAME',
+        pwd: '$MONGO_INITDB_ROOT_PASSWORD',
+        roles: [ { role: 'readWrite', db: '$MONGO_INITDB_DATABASE' } ]
+    })
```
Comment (jakubkoci) [inline .docker/mongodb/initdb.d/usercreate.sh:8]: What if we extract this JS too, to keep it aligned with the replica set init?

File: packages/frontend/src/common/components/GenericErrorContents.tsx [hunk 6]
```diff
@@ -0,0 +1,19 @@
...
+
+export function GenericErrorContents() {
+  return (
+    <div className="flex flex-col items-center gap-8 text-center">
+      <Failure />
+      <div className="flex flex-col gap-2">
+        <Text2xlBold>Oops, something went wrong</Text2xlBold>
+        <TextSm>
+          Try to refresh the page. If the issue persists, contact us at
+          lorem@ipsum.com
```
Comment (jakubkoci) [inline packages/frontend/src/common/components/GenericErrorContents.tsx:13]:
```suggestion
          ssi@absa.africa
```

File: .docker/mongodb/scripts/rs-init.js
Comment (niall-shaw) [reply .docker/mongodb/scripts/rs-init.js:1]: Replica Set.

https://www.mongodb.com/docs/manual/replication/

Comment (jakubkoci) [reply .docker/mongodb/scripts/rs-init.js:1]: Then name it properly ;)

File: packages/frontend/src/common/hooks.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (jakubkoci) [reply packages/frontend/src/common/hooks.ts:5]: Isn't the name `ServerError` misleading? What does it mean? Is it that the error happened on the server-side? Does it mean that all other errors happen only on the client-side?

[the remark this answers is not in the document]
Comment (jakubkoci) [reply packages/frontend/src/common/hooks.ts:5]: Honestly, I don't see much benefits in using custom hook, it seems that only difference is the following, right?

```ts
type ServerError = { server?: never }
useForm<Inputs & ServerError>
```

```ts
useFormWithServerError()
```

[the remark this answers is not in the document]
Comment (jakubkoci) [reply packages/frontend/src/common/hooks.ts:5]: Yes, probably just having `ServerError` or `GenericError` should be good enough.

**Review by jakubkoci — APPROVED** [review]

**Review by jakubkoci — APPROVED** [review]

**Review by jakubkoci — APPROVED** [review]

**Pull request merged**